### PR TITLE
Complete find meeting time query

### PR DESCRIPTION
**Pull request by dalisc**

### CODE REVIEW

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,84 @@
...
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) { 
+    // check edge cases before calling helper functions
+    if (request.getDuration() == 0 || request.getDuration() > 1440 ) {
```
Comment (123cristen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: What does `1440` represent? For constants it's useful to pull them out with a descriptive variable name, I.e. `const int MINUTES_IN_A_DAY = 1440;`

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,84 @@
...
+  * A time range isn't available if there is an event happening and some people in the
+  * event are required to be in the meeting.
+  * If there are no unavailable slots, this function returns an empty arraylist.
+  *
+  * @param events the existing events with attendees that already planned
+  * @param request the meeting request containing the duration and attendees needed
+  * @return an arraylist containing TimeRanges that aren't available for the meeting, sorted by start time
+  */
+  private List<TimeRange> getUnavailableTimeRanges(Collection<Event> events, MeetingRequest request) {
+    List<TimeRange> unavailableTimeRanges = new ArrayList<>();
```
Comment (123cristen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: You can simplify this using the Java stream library:
```
events.stream()
.filter(e -> !Collections.disjoint(e.getAttendees(), request.getAttendees())
.map(e -> e.when())
.sorted(TimeRange.ORDER_BY_START)
.collect(Collectors.toList());`

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,84 @@
...
+
+    // sort the time ranges in chronological order
+    Collections.sort(unavailableTimeRanges, TimeRange.ORDER_BY_START);
+
+    return unavailableTimeRanges;
+  }
+
+  /**
+  * Returns the time ranges for which the meeting request can happen in. 
+  * This returns the "inverse" of the unavailable timeslots.
```
Comment (123cristen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: It's not exactly the inverse, right? Maybe add a note that it only adds time slots with a duration greater than the required duration

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,84 @@
...
+
+        // the start time of the next available time range is the end time of the current time range
+        // or the end time of the time range in which this current time range is a subset of
+        if (tr.end() > start) {
+            start = tr.end();
+        }
+
+    }
+
+    int endOfDay = 1440;
```
Comment (123cristen) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This is equivalent to `TimeRange.END_OF_DAY + 1`, or just include it as a const above.

**Review by 123cristen — CHANGES_REQUESTED** [review]
Nice! Very cleanly written :)

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (dalisc) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: thanks! this is cool

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (dalisc) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: i'll do this instead

**Review by 123cristen — APPROVED** [review]
I'll approve this now but note that there's still a few comments to be addressed :)

**Comment by dalisc** [comment]
> I'll approve this now but note that there's still a few comments to be addressed :)

**Pull request merged**